llvm/lib/Target/DirectX/DXIL.td

```diff
@@ -13,353 +13,777 @@
+def tempRegStore : DXILOpClass;
+def tertiary : DXILOpClass;
+def texture2DMSGetSamplePosition : DXILOpClass;
+def textureGather : DXILOpClass;
+def textureGatherCmp : DXILOpClass;
+def textureGatherRaw : DXILOpClass;
+def textureLoad : DXILOpClass;
+def textureStore : DXILOpClass;
+def textureStoreSample : DXILOpClass;
+def threadId : DXILOpClass;
+def threadIdInGroup : DXILOpClass;
+def traceRay : DXILOpClass;
+def unary : DXILOpClass;
+def unaryBits : DXILOpClass;
+def unpack4x8 : DXILOpClass;
+def viewID : DXILOpClass;
+def waveActiveAllEqual : DXILOpClass;
+def waveActiveBallot : DXILOpClass;
+def waveActiveBit : DXILOpClass;
+def waveActiveOp : DXILOpClass;
+def waveAllOp : DXILOpClass;
+def waveAllTrue : DXILOpClass;
+def waveAnyTrue : DXILOpClass;
+def waveGetLaneCount : DXILOpClass;
+def waveGetLaneIndex : DXILOpClass;
+def waveIsFirstLane : DXILOpClass;
+def waveMatch : DXILOpClass;
+def waveMatrix_Accumulate : DXILOpClass;
+def waveMatrix_Annotate : DXILOpClass;
+def waveMatrix_Depth : DXILOpClass;
+def waveMatrix_Fill : DXILOpClass;
+def waveMatrix_LoadGroupShared : DXILOpClass;
+def waveMatrix_LoadRawBuf : DXILOpClass;
+def waveMatrix_Multiply : DXILOpClass;
+def waveMatrix_ScalarOp : DXILOpClass;
+def waveMatrix_StoreGroupShared : DXILOpClass;
+def waveMatrix_StoreRawBuf : DXILOpClass;
+def waveMultiPrefixBitCount : DXILOpClass;
+def waveMultiPrefixOp : DXILOpClass;
+def wavePrefixOp : DXILOpClass;
+def waveReadLaneAt : DXILOpClass;
+def waveReadLaneFirst : DXILOpClass;
+def worldRayDirection : DXILOpClass;
+def worldRayOrigin : DXILOpClass;
+def worldToObject : DXILOpClass;
+def writeSamplerFeedback : DXILOpClass;
+def writeSamplerFeedbackBias : DXILOpClass;
+def writeSamplerFeedbackGrad : DXILOpClass;
+def writeSamplerFeedbackLevel : DXILOpClass;
+
+// This is a sentinel definition. Hence placed at the end here
+// and not as part of the above alphabetically sorted valid definitions.
+// It is never used to construct the name of DXIL Op call name.
+// Additionally it is capitalized unlike all the others.
+def UnknownOpClass : DXILOpClass;
+
+// Shader stages
+class ShaderStage;
+
+def compute : ShaderStage;
+def domain : ShaderStage;
+def hull : ShaderStage;
+def pixel : ShaderStage;
+def vertex : ShaderStage;
+def geometry : ShaderStage;
+def library : ShaderStage;
+def amplification : ShaderStage;
+def mesh : ShaderStage;
+def node : ShaderStage;
+def raygeneration : ShaderStage;
+def intersection : ShaderStage;
+def anyhit : ShaderStage;
+def closesthit : ShaderStage;
+def callable : ShaderStage;
+def miss : ShaderStage;
+
+// Primitive predicate
+class Pred;
+
+// Following classes represent predicate that evaluate to the payload,
+// if one xists, unconditionally.
+
+// DXIL Op attributes defined as true predicates, if specified
+
+def ReadOnly : Pred;
+def ReadNone : Pred;
+def IsDerivative : Pred;
+def IsGradient : Pred;
+def IsFeedback : Pred;
+def IsWave : Pred;
+def NeedsUniformInputs : Pred;
+def IsBarrier : Pred;
```

Comment:
I'm probably missing something, but how come these attributes are predicates?  Can they be used in a constraint?

Reply:
> I'm probably missing something, but how come these attributes are predicates? Can they be used in a constraint?

Yes. It would express constraints when the attribute is true. Not yet seen a need for such a specification, though. The idea is to use a unified TableGen class to model an attribute as a predicate that is always true instead of using a separate `class Attribute` to convey the same semantics.

Reply:
I'm afraid I don't follow what "constraints when the attribute is true" means.  If these attributes are always true, then what would a constraint involving one of these attributes mean?

Reply:
As an example, consider the hypothetical example where an attribute represented as a predicate can be used in constraint (per the original question) - `Constraint<IsWave, [SomePredicate, SomeOtherPredicate]>`. It represents that if `IsWave` is true, the predicates `SomePredicate` and `SomeOtherPredicate` are true for the DXIL Operation. This is similar to the specification `Constraints<SMVersion<SM6_0>, [Overloads<[llvm_half_ty, llvm_float_ty, llvm_double_ty]>]>]` that represents that if the predicate `SMVersion<SM6_0>` is true, the predicate `Overloads<llvm_half_ty, llvm_float_ty, llvm_double_ty]>]>]` is true.

Reply:
So my confusion here is that it seems that `IsWave` is defined to be always true (line 251 above).  How could it be not true?

Reply:
A predicate - unconditional or conditional - when not specified, represents that it is not true for that operation. For example, if `IsWave` (unconditional) is specified, the operation has the property of being a wave operation. Absence of it indicates that it is a not constrained to be a wave operation. Similarly, `Constraints<SMVersion<SM6_0>, [P1, P2]>` (conditional predicate) indicates that the operation has the properties `P1` and `P2` for the predicate `SMVersion<SM<6_0>` (i.e., `SMVersion<SM6_0>` is true, by virtue of it being specified). Absence of any `Constraints<...>` indicates that the operation has no conditional predicates (i.e., properties) and is valid in all circumstances.

Reply:
I think some confusion here is that this PR is now diverging quite a from the design described in DXILOpTableGenDesign - where there was a DXILAttributes class.

If we're going to be implementing a new design, maybe we need to go back and get that spec updated first so we can be sure we're all on the same page.

Reply:
> I think some confusion here is that this PR is now diverging quite a from the design described in DXILOpTableGenDesign - where there was a DXILAttributes class.
> 
> If we're going to be implementing a new design, maybe we need to go back and get that spec updated first so we can be sure we're all on the same page.

Pushed changes in accordance with the latest design doc.